File: practice_mvvm/LiveData/app/src/main/java/com/miraclehwan/livedata/MainActivity.java
package com.miraclehwan.livedata;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;

import com.miraclehwan.livedata.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    private MainViewModel mainViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        mainViewModel = ViewModelProviders.of(this).get(MainViewModel.class);
        binding.setMainViewModel(mainViewModel);

        mainViewModel.getCurrentName().observe(this, new Observer<String>() {
            @Override
            public void onChanged(@Nullable String s) {
                binding.textview.setText(s);
            }
        });

        getSupportFragmentManager().beginTransaction()
                .replace(R.id.frame_layout, MainFragment.newInstance())
                .commit();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.e("miraclehwan", "onDestroy");
    }
}

File: TestMVVM/app/src/main/java/com/miracle/testmvvm/main/MainActivity.kt
package com.miracle.testmvvm.main

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.miracle.testmvvm.R
import com.miracle.testmvvm.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    val mainViewModel: MainViewModel by lazy {
        ViewModelProvider(this@MainActivity)[MainViewModel::class.java]
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main).apply {
            vm = mainViewModel
        }

        initObserver()
    }

    fun initObserver() {
        mainViewModel.compositeDisposable.add(
            mainViewModel.intentSubject
                .subscribe { intentInfo ->
                    startActivity(Intent(this, intentInfo.className).apply {
                        putExtras(intentInfo.bundle)
                    })
                })
    }
}
